lib/afinter.c

```diff
@@ -496,3 +501,83 @@ afinter_global_deinit(void)
     }
   current_internal_source = NULL;
 }
+
+/****************************************************************************
+ * Live collection of internal log messages.
+ ****************************************************************************/
+
+AFInterLive
+afinter_start_live_collection(void)
+{
+  AFInterSource *self;
+
+  if (is_live_collection)
+    return AFINTER_LIVE_COLLECTION_RUNNING;
+
+  if (current_internal_source != NULL)
+    return AFINTER_INTERNAL_SRC_PRESENT;
+
+  self = g_new0(AFInterSource, 1);
+
+  g_static_mutex_lock(&internal_msg_lock);
+  current_internal_source = self;
+  g_static_mutex_unlock(&internal_msg_lock);
+
+  is_live_collection = TRUE;
+
+  return AFINTER_LIVE_COLLECTION_STARTED;
+}
+
+void
+afinter_reset_live_collection(void)
+{
+  afinter_stop_live_collection();
+
+  if (is_live_collection)
+    {
+      if (internal_msg_queue)
+        _release_internal_msg_queue();
+    }
+}
+
+void
+afinter_stop_live_collection(void)
+{
+  if (is_live_collection)
+    {
+      if (current_internal_source != NULL)
+        {
+          AFInterSource *self = current_internal_source;
+
+          g_static_mutex_lock(&internal_msg_lock);
+          current_internal_source = NULL;
+          g_static_mutex_unlock(&internal_msg_lock);
+
+          is_live_collection = FALSE;
+          g_free(self);
+        }
+    }
+}
+
+void
+afinter_get_collected_messages(GString *result)
+{
+  LogMessage *msg;
+  const gchar *msg_text;
+
+  g_static_mutex_lock(&internal_msg_lock);
+
+  while (!g_queue_is_empty(internal_msg_queue))
```

Comment:
I steadily get deadlock in the following case:
```
@version: 3.25

log {
    source { example-msg-generator(); };
    destination { file(/dev/stdout); };
};
```

Start syslog-ng with `./syslog-ng -F -f ../etc/internal.conf` (! just -F, not -Fdt. With -Fdt, I did not see deadlock yet).

After start and stop, syslog-ng hangs. I crashed syslog-ng with segv (kill -segv <pid>), then checked the core dump with gdb: I see syslog-ng hangs while trying to grab a lock during pushing internal message. Moving up in the stack, the message is triggered by `g_queue_is_empty` (through a `g_log` call). The hint is `internal_msg_queue` is empty for some reason, and I think the code wants to emit the same assert that I mentioned in the other comment. So maybe the root cause and the fix is the same.

Reply:
@furiel : this is fixed in the new changes.